pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java

```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
                 return;
             }
             execState = ExecState.RUNNING;
-            final String line = words.stream().collect(Collectors.joining(" "));
-            if (StringUtils.isBlank(line)) {
+            final String inputLine = words.stream().collect(Collectors.joining(" "));
+            if (StringUtils.isBlank(inputLine)) {
                 continue;
             }
-            if (isQuitCommand(line)) {
+            if (isQuitCommand(inputLine)) {
+                if (shellMode != ShellMode.DEFAULT){
+                    output(String.format("Exiting from %s shell mode", shellMode.command), terminal);
+                    promptMessage = promptMessage.substring(0, promptMessage.lastIndexOf(shellMode.command) - 1);
+                    prompt = createPrompt(promptMessage);
+                    shellMode = ShellMode.DEFAULT;
+                    continue;
+                }
                 exit(0);
                 return;
             }
+            if (shellMode != ShellMode.DEFAULT) {
+                words.add(0, shellMode.command);
+            }
+            final String line = words.stream().collect(Collectors.joining(" "));
             if (isHelp(line)) {
                 shellCommander.usage(System.out);
                 continue;
             }
+            ShellMode newShellMode = ShellMode.valueOfCommand(line.toLowerCase(Locale.ROOT));
+            if (newShellMode != null){
+                shellMode = newShellMode;
+                promptMessage = promptMessage.concat(" ").concat(shellMode.command);
+                prompt = createPrompt(promptMessage);
```

Comment:
See the previous comment about not modifying promptMessage. In Java, the compiler will optimize String concatenation. You can simply use `+` to concatenate Strings.
```suggestion
                prompt = createPrompt(promptMessage + " " + shellMode.command));
```